app2/src/main/java/com/example/recipebox/DatabaseHelper.kt

```diff
@@ -0,0 +1,110 @@
+package com.example.recipebox
+
+import android.database.sqlite.SQLiteOpenHelper
+import android.database.sqlite.SQLiteDatabase
+import android.content.ContentValues
+import android.content.Context
+import kotlin.String
+import java.util.ArrayList
+
+class DatabaseHelper(context: Context?) :
+    SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {
+    override fun onCreate(db: SQLiteDatabase) {
+        val CREATE_CONTACTS_TABLE =
+            "create table $TABLE_CONTACTS($KEY_ID INTEGER PRIMARY KEY,$KEY_NAME TEXT,$KEY_PHONE TEXT)"
+        db.execSQL(CREATE_CONTACTS_TABLE)
+    }
+
+    override fun onUpgrade(db: SQLiteDatabase, i: Int, il: Int) {
+        db.execSQL("DROP TABLE IF EXISTS $TABLE_CONTACTS")
+        onCreate(db)
+    }
+
+    fun addContact(contact: Contact) {
+        val db = this.writableDatabase
+        val values = ContentValues()
+        values.put(KEY_NAME, contact.name)
+        values.put(KEY_PHONE, contact.phone_number)
+        db.insert(TABLE_CONTACTS, null, values)
+        db.close()
+    }
+
+    fun getContact(id: Int): Contact {
+        val db = this.readableDatabase
+        val cursor = db.query(
+            TABLE_CONTACTS,
+            arrayOf(
+                KEY_ID,
+                KEY_NAME,
+                KEY_PHONE
+            ),
+            "$KEY_ID=?",
+            arrayOf(id.toString()),
+            null,
+            null,
+            null,
+            null
+        )
+        try {
+            cursor?.moveToFirst()
+            return Contact(
+                cursor!!.getString(0).toInt(),
+                cursor.getString(1),
+                cursor.getString(2)
+            )
+        } finally {
+            cursor.close()
+        }
+    }
+
+    val allContacts: List<Contact>
+        get() {
+            val contactList: MutableList<Contact> = ArrayList()
+            val selectQuery = "SELECT *FROM $TABLE_CONTACTS"
+            val db = this.writableDatabase
+            val cursor = db.rawQuery(selectQuery, null)
+            if (cursor.moveToFirst()) {
+                do {
+                    val contact = Contact()
+                    contact.id = cursor.getString(0).toInt()
+                    contact.name = cursor.getString(1)
+                    contact.phone_number = cursor.getString(2)
+                    contactList.add(contact)
+                } while (cursor.moveToNext())
+            }
+            cursor.close()
```

Comment:
Always make sure to close your cursors, or you could hit an open cursor limit.

Reply:
The cursors were not closed at the end in the original Java file before I converted to Kotlin

Reply:
Good catch.